padertorch/io.py

```diff
@@ -79,3 +80,103 @@ def get_new_storage_dir(
     basedir = Path(os.environ['STORAGE_ROOT']) / experiment_name
     del experiment_name
     return get_new_subdir(**locals())
+
+
+def dump_config(
+        config,
+        path,
+):
+    """
+    This functions dumps a config as json or yaml.
+
+     - Convert factory values to str if necessary
+     - Move factory to be the first key in the dict
+     - Sort the order of the kwargs based on the signature of the factory.
+
+    """
+    path = Path(path)
+    path.write_text(dumps_config(config, path.suffix))
+
+
+def dumps_config(
+        config,
+        format='.yaml'  # '.json' or '.yaml'
+):
+    """
+    >>> print(dumps_config({'a': 1, 'b': 2}, '.yaml'))
+    a: 1
+    b: 2
+    <BLANKLINE>
+    >>> print(dumps_config({'b': 2, 'a': 1}, '.yaml'))
+    b: 2
+    a: 1
+    <BLANKLINE>
+    >>> print(dumps_config({'a': 1, 'b': 2}, '.json'))
+    {
+      "a": 1,
+      "b": 2
+    }
+    >>> print(dumps_config({'b': 2, 'a': 1}, '.json'))
+    {
+      "b": 2,
+      "a": 1
+    }
+
+    >>> import torch
+    >>> print(dumps_config({'factory': torch.nn.ReLU, 'b': 2}, '.yaml'))
+    factory: torch.nn.modules.activation.ReLU
+    b: 2
+    <BLANKLINE>
+    >>> print(dumps_config({'b': 2, 'factory': torch.nn.ReLU}, '.yaml'))
+    factory: torch.nn.modules.activation.ReLU
+    b: 2
+    <BLANKLINE>
+
+    """
+    import padertorch as pt
+    config = pt.configurable.recursive_class_to_str(config, sort=True)
+    if format == ".json":
+        from paderbox.io import dumps_json
+        return dumps_json(config, sort_keys=False)
+    elif format == ".yaml":
+        from paderbox.io.yaml_module import dumps_yaml
+        return dumps_yaml(config, sort_keys=False)
+    else:
+        raise NotImplementedError(format)
+
+
+def load_config(
+        path
+):
+    import paderbox as pb
+    return pb.io.load(path)
+
+
+def loads_config(
+        content,
+        format=None,  # '.yaml' or '.json'
+):
```

Comment:
this should be the same default as above in dumps_config

Reply:
The default is compatible to `dumps_config`, so I thought this is fine.

Reply:
but why take None here and '.yaml' for dumps_config?

Reply:
Since we have a limited number of supported serialization formats and the outer type is a dict, we can simply guess the format (i.e. json starts with `{`, while yaml doesn't.)

Reply:
Shouldn't the default for ```loads_config``` still be None and load_config does the guessing?

Reply:
But sure I see your point. It just looks a little bit strange to have different defaults. But its fine

Reply:
In `load_config` it is not necessary to guess, because the suffix tells which format it is.
I just wrote the doctests and it was annoying to specify the format, so I added the guessing to `loads_config`.

If it is to much magic, I can remove it (I expect, not end user will use `loads` and `dumps`).

Reply:
I removed not the auto-detection